build.sbt

```diff
@@ -622,6 +625,30 @@ lazy val tests = projectMatrix
   )
   .http4sPlatform(allJvmScalaVersions, jvmDimSettings)
 
+lazy val weaverTests = projectMatrix
+  .in(file("modules/weaver-tests"))
```

Comment:
let's rename it to something else. The fact that we're using weaver is a detail, and the module doesn't aim at providing facilities for users of smithy4s to test their code in the context of the weaver framework.   

It could be `compliance` or something of the sort

Reply:
At some point I think it'd be pretty great to provide test trait support that'd be test-framework-agnostic, too. Doesn't have to be in the initial appearance though

Reply:
The tedious thing is that there aren't any standard assertions. It means you have to come up with some data structure that does roughly the same thing as test-framework assertions. It's going be more boilerplaty (you have to create custom failure messages).

@daddykotex, if you're up for a little challenge, I think we should actually try and pursue it in this PR. It'd mean that you'd be producing something like a `List[ComplianceTest[F]]` that the framework of choice could run however it pleases . 

```
case class ComplianceTest[F[_]](name: String, run: F[ComplianceResult]) 
```

Reply:
Ahh, ok, I did not think we were looking for something like that

Basically, a user would:

1. add dependency on this module: `smithy4s-compliance-tests`
2. add a depeendency on their model (smithy stuff)
3. the user can then generate a `List[ComplianceTest]` that they can register with the testing framework they want: (munit, weaver, w/e)

Right now though, we're only testing smithy4s stuff. What I mean is that we use a smithy4s generated http4s client along with a smithy4s generated http4s route and we _wire_ together to run assertion. That's great, but I'd argue it's not that useful to a user: we test the generated code.

I think that a more useful compliance test would allow the user to wire up an external server (maybe something not implemented with smithy4s) and/or an external client.

Does that make any sense?

Reply:
> I think that a more useful compliance test would allow the user to wire up an external server (maybe something not implemented with smithy4s) and/or an external client. 

> What I mean is that we use a smithy4s generated http4s client along with a smithy4s generated http4s route and we wire together to run assertion.

Well even if you're technically correct in that we're only testing smithy4s-derived (not-generated) constructs, the same tests can be used to test against several http protocols and backends, which might be implemented by third parties. It doesn't cost us much to lose the weaver dependency on a published artifact. As a matter of fact, despite being the maintainer of weaver, I feel more comfortable not publishing an artifact that has a dependency on weaver. 

So even if we are likely the only ones using this artifact, I'd much rather treat it like any other published artifact we are issuing in this library and prevent tight-coupling when not necessary. This philosophy has paid off handsomely so far and has avoided a large number of headaches.

Reply:
my comment did not mean to imply that dropping weaver was not a good idea. I'll do it, this whole suggestion make sense, I meant something else:

currently, we test http4s client with http4s router, no true server  to client communication. should we expose a way for user to spawn the clients test but to target their own server located at for example http://localhost:8080

(basically, a way for the user to target an existing server (or the opposite, to spawn a server that they can hit with their client))

Reply:
Oh right, sorry, I see what you mean. 

We should copy what is done in PizzaSpec/PizzaClientSpec. Namely, [this entrypoint](https://github.com/disneystreaming/smithy4s/pull/424/files#diff-09b109d2d2c3403953cd05ef5d2c8966037aa476f4eb05d8caaea7c6b00e9577R62-R69) should be replace by something that takes 

``` 
client: Either[HttpApp[IO] => Resource[IO, smithy4s.Monadic[Alg, F]], Int => Resource[IO,smithy4s.Monadic[Alg, IO]]], 
server: smithy4s.Monadic[Alg, F] => Resource[F, (Client[F], Uri)] 
```

In other words, on the client side, it is : 

"give me either a function that takes HttpApp or a port and returns a Algebra, or a function that takes a port and returns an algebra, and I'll be able to spin a server to test your client"

On the server side, it is : "give me a function that, provided an algebra, spins a server and returns me with a client and the URI at which the server is running, and I'll use that client to test your server". 

This would be sufficient for client tests to target existing servers. Regarding server tests, we can obviously not test the implementation of the server-side of the protocol without controlling the business logic of server in question.

Reply:
I think this is what is sort of blocking me. Let's focus on the **client request tests**.

Looking at the current implementation of the PizzaClientSpec :

```scala
def makeClient: Either[
    HttpApp[IO] => Resource[IO, PizzaAdminService[IO]],
    Int => Resource[IO, PizzaAdminService[IO]]
  ]
```
I went ahead and did most of the work using that, but then, when it came to it, I tried to wire it all together: but I cant implement the right side of this function. Whats the body of the following:

```scala
object Http4sPizzaClientSpec2 extends smithy4s.tests.PizzaClientSpec {
  def makeClient: Either[
    HttpApp[IO] => Resource[IO, PizzaAdminService[IO]],
    Int => Resource[IO, PizzaAdminService[IO]]
  ] = Right { port =>
    ???
  }
}
```

I have a feeling I'll feel very dumb in about just a minute, but that's okay :D

Reply:
@daddykotex the Right-hand-side is there for testing clients that are not backed by http4s. The port lets the implementor formulate a `localhost:$port` URI and that to some function, alongside an http-client, that will create a smithy4s instance. 

When the implementor provides a `Right`, the spec should instantiate an http-server at a random port, then use the lambda contained by the `Right` to get a client that will exercise that http-server. 

Hope that makes sense.

Reply:
oh wow, ok, that's not how I was understanding it
I'll make more progress on that tomorrow morning

that makes a lot of sense

------

edit: wrote that last night, forgot to send :D